File: v3_open_close/python/notification_service1.py
class EmailService:

    def send_notification(self, info):
        print(f'Email: {info}')

class MessageService:

    def send_notification(self, info):
        print(f'Message: {info}')

class NotificationService1:

    def send_notification(self, service_type, info):
        if service_type == 'email':
            self.send_email(info)
        elif service_type == 'message':
            self.send_message(info)
    
    def send_email(self, info):
        es = EmailService()
        es.send_notification(info)
    
    def send_message(self, info):
        ms = MessageService()
        ms.send_notification(info)

ns = NotificationService1()
ns.send_notification('email', 'hello')
ns.send_notification('message', 'hello')

File: v3_open_close/java/NotificationService1.java
public class NotificationService1 {

    public static void main(String[] args) {
        NotificationService1 ns = new NotificationService1();
        ns.sendNotification("email", "helo");
        ns.sendNotification("message", "hello");
    }

    public void sendNotification(String serviceType, String info) {
        if (serviceType == "email") {
            sendEmail(info);
        } else if (serviceType == "message") {
            sendMessage(info);
        }
    }

    public void sendEmail(String info) {
        EmailService es = new EmailService();
        es.sendNotification(info);
    }

    public void sendMessage(String info) {
        MessageService ms = new MessageService();
        ms.sendNotification(info);
    }
}

class EmailService {
    public void sendNotification(String info) {
        System.out.printf("Email: %s\n", info);
    }
}

class MessageService {
    public void sendNotification(String info) {
        System.out.printf("Message: %s\n", info);
    }
}
